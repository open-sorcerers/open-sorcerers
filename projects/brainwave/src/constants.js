export const BP = "basePath"
export const BR = "brain"
export const CO = "content" // comes from cosmiconfig
export const CC = "config" // comes from cosmiconfig
export const CF = "configFile"
export const FI = "filepath"
export const FT = "fileTypes"
export const MC = "control"
export const NS = "namespace"
export const RT = "root"
export const GLOB = "glob"
export const TK = "telepathy"
export const ST = "stats"
export const DA = "data"
export const DR = "dryRun"
export const FC = "fileContent"
export const RP = "relativePath"

export const INIT = `init`
