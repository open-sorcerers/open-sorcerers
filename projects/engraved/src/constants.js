export const C = {
  n: "\n"
}

export const REGEX_HEX3 = /[a-f0-9]{3}/
export const REGEX_HEX6 = /[a-f0-9]{6}/
