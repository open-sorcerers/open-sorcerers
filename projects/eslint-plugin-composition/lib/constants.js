const PARENT = 'parent';
const OBJECT = 'object';
const PROPERTY = 'property';
const NAME = 'name';
const TYPE = 'type';
const PROPS = 'props';

module.exports = {
  NAME,
  OBJECT,
  PARENT,
  PROPERTY,
  PROPS,
  TYPE
};
