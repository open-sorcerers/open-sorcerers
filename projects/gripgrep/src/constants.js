export const KEYS = Object.freeze({
  CWD: "cwd",
  REGEX: "regex",
  GLOBS: "globs",
  STRING: "string"
})

export const NIL = `nil`
