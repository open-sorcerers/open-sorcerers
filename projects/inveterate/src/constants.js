export const DEFAULT_CONFIG = { scheduled: false }

export const EVERY = Object.freeze({
  MINUTE: "* * * * *",
  FIVE_MINUTES: "*/5 * * * *",
  FIFTEEN_MINUTES: "*/15 * * * *",
  THIRTY_MINUTES: "*/30 * * * *",
  HOUR: "0 * * * *",
  MONTH: "0 0 1 * *",
  WEEKDAY_START: "0 9 * * 1-5",
  WEEKDAY_END: "0 5 * * 1-5",
  NIGHT: "0 0 * * *",
  WEEKEND_START: "0 0 * * 0",
  WEEKEND_END: "23 0 * * 6"
})

