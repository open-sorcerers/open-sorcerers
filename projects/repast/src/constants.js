import { map, pipe, reduce, mergeRight } from "ramda"
/* eslint-disable string-literal/no-string-literal */

export const REGEXES = {
  startsWithParens: /^\(/,
  endsWithParens: /\)$/
}

export const TYPES = {
  HM_TYPE: "RepastData",
  HM_SIGNATURE: "RepastSignature"
}

export const CHARACTERS = Object.freeze({
  __of__: `âˆ‹`,
  newline: "\n",
  comment: "//",
  space: " ",
  openParens: "(",
  closeParens: ")",
  asterisk: "*",
  empty: "",
  space: " ",
  doubleColon: "::",
  equal: "="
})
export const C = CHARACTERS

export const KEYWORDS = Object.freeze({
  REPAST: "@repast"
})
export const K = KEYWORDS

const makeObjectFromStrings = pipe(
  reduce((xx, yy) => mergeRight(xx, { [yy]: yy }), {}),
  Object.freeze
)

// read !cat file.js | grep 'L\..'
export const LITERALS = makeObjectFromStrings([
  "Boolean",
  "Global",
  "Identifier",
  "String",
  "ast",
  "arity",
  "body",
  "globals",
  "input",
  "json",
  "loc",
  "name",
  "nil",
  "object",
  "output",
  "params",
  "program",
  "signature",
  "string",
  "type",
  "undefined",
  "utf8",
  "value"
])

export const L = LITERALS

export const FLAGS = makeObjectFromStrings([L.input, L.output, L.json])
export const nf = map(z => "--" + z)

export const PRIMITIVE_TYPES = makeObjectFromStrings([
  "Boolean",
  "String",
  "Number",
  "Void",
  "Object",
  "Array",
]);

export const GLOBALS = pipe(
  reduce((agg, thing) => mergeRight(agg, { [thing.name]: thing }), {})
)([
  Array,
  Boolean,
  Date,
  Error,
  Function,
  Number,
  Object,
  Promise,
  Proxy,
  RegExp,
  String,
  Symbol
])
