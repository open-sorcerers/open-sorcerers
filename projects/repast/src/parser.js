import {
  __,
  all,
  ap,
  assoc,
  concat,
  curry,
  dissoc,
  equals,
  filter,
  identity as I,
  ifElse,
  indexOf,
  init,
  last,
  map,
  mergeRight,
  nth,
  objOf,
  of,
  pathOr,
  pipe,
  prop,
  propEq,
  reduce,
  slice,
  split,
  splitWhen,
  startsWith,
  uniq,
  when
} from "ramda"
import { parse } from "recast"
import { trace } from "xtrace"
import { C, GLOBALS, L, K, REGEXES, TYPES } from "./constants"

// @repast Config = { input: String, output: String, json: Boolean }
// @repast NestedSignature = { nested: Boolean, value: Array, lastWasEnder: Boolean }
// @repast RepastDefinition = { type: String, name: String, definition: [String] }
// @repast RepastSignature = { type: String, name: String, signature: [String], globals: [Function] }
// @repast RepastParse = { definitions: [RepastDefinitions], signatures: [RepastSignature] }

// @repast box :: any -> [any]
const box = x => (Array.isArray(x) ? x : [x])

// @repast readNestedSignatures :: [String] -> NestedSignature
const readNestedSignatures = reduce(
  // @repast _readNestedSignatureReducer :: NestedSignature -> String -> NestedSignature
  function _readNestedSignatureReducer(agg, xx) {
    const starter = xx.startsWith(C.openParens)
    const ender = xx.endsWith(C.closeParens)
    const { nested, value, lastWasEnder } = agg
    const sliced = { start: init(value), end: last(value) }

    const yy = xx
      .replace(REGEXES.startsWithParens, C.empty)
      .replace(REGEXES.endsWithParens, C.empty)
    return {
      lastWasEnder: ender,
      value: starter
        ? concat(value, box(yy))
        : nested
        ? concat(box(init(value)), [concat(box(last(value)), box(yy))])
        : concat(value, [yy]),
      nested: ender ? false : nested || starter ? true : nested
    }
  },
  { nested: false, value: [], lastWasEnder: false }
)

// @repast j2 :: any -> String
const j2 = x => JSON.stringify(x, null, 2)

// @repast hasString :: String -> String -> Boolean
const hasString = curry((a, b) => indexOf(a, b) > -1)

// @repast trim :: String -> String
const trim = x => x.trim()

// @repast resolveDataDefinitions :: AST -> RepastDefinition
const resolveDataDefinitions = pipe(
  of,
  ap([nth(1), slice(3, Infinity)]),
  ([name, definition]) => {
    const def = JSON.parse(definition.join(" "))
    return { name, definition: def, type: HM.TYPE }
  }
)

// @repast resolveBlockComments :: [String] -> [String]
const resolveBlockComments = map(z =>
  z[0] === C.asterisk ? z.slice(1, Infinity) : z
)

// @repast getGlobalsFromSignature :: Config -> [String]
const getGlobalsFromSignature = curry((config, sig) =>
  pipe(
    filter(z => !!GLOBALS[z]),
    config.json ? map(z => GLOBALS[z]) : I,
    uniq
  )(sig)
)
// @repast isRepastDataDefinition :: [String] -> Boolean
const isRepastDataDefinition = x => x[0] === K.REPAST && x[1] === C.equal

// @repast matchesExpectedPattern :: [String] -> Boolean
const matchesExpectedPattern = pipe(
  of,
  ap([pipe(nth(0), equals(K.REPAST)), pipe(nth(2), equals(C.doubleColon))]),
  all(x => x)
)

// @repast parseSignatureGivenConfig :: Config -> [String] -> RepastSignature
const parseSignatureGivenConfig = curry((config, x) => {
  const sig = x.slice(3, Infinity)
  return {
    [L.type]: TYPES.HM_SIGNATURE,
    [L.name]: nth(1, x),
    [L.signature]: sig,
    [L.globals]: getGlobalsFromSignature(config, sig)
  }
})

// @repast trimLines :: [String] -> [String]
const trimLines = map(trim)

// @repast trimLines :: [String] -> [String]
const removeEmptyLines = filter(Boolean)

// @repast hindleymilnerize :: Config -> String -> [RepastSignature|RepastDefinition]
const hindleymilnerize = curry((config, str) =>
  pipe(
    split(C.space),
    trimLines,
    resolveBlockComments,
    removeEmptyLines,
    readNestedSignatures,
    prop(L.value),
    ifElse(
      isRepastDataDefinition,
      resolveDataDefinitions,
      ifElse(
        matchesExpectedPattern,
        parseSignatureGivenConfig(config),
        trace("This is not a correctly formatted comment?")
      )
    )
  )(str)
)

// @repast grabEntitiesWithMagicComments :: Config -> [String] -> RepastParse
const grabEntitiesWithMagicComments = curry((config, list) =>
  reduce((agg, ast) => {
    const { comments } = ast
    if (
      comments.length &&
      filter(pipe(prop(L.value), hasString(K.REPAST)), comments).length
    ) {
      return pipe(
        map(prop(L.value)),
        filter(hasString(K.REPAST)),
        map(comment => mergeRight(hindleymilnerize(config, comment), { ast })),
        concat(agg)
      )(comments)
    }
    return agg
  }, [])(list)
)

// @hm conditionallyConvertToJSON :: Config -> [AST] -> [AST]|JSON
const conditionallyConvertToJSON = curry((config, xx) =>
  when(() => prop(L.json, config), map(pipe(dissoc(L.ast), j2)))(xx)
)

// @hm parseWithConfig :: Object -> String -> Array
export const parseWithConfig = curry((config, str) =>
  pipe(
    parse,
    pathOr([], [L.program, L.body]),
    grabEntitiesWithMagicComments(config),
    map(when(propEq(L.type, TYPES.HM_TYPE), trace("this is a thing"))),
    conditionallyConvertToJSON(config)
  )(str)
)
