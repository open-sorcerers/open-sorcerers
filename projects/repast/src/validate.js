import {
  always,
  append,
  curry,
  dropLast,
  ifElse,
  last,
  length,
  map,
  pipe,
  propEq,
  reduce,
  reject,
} from "ramda";
import { print } from "recast";
import F from "fluture";
import vm from "vm";

import { PRIMITIVE_TYPES } from "./constants";

const parseSignature = pipe(
  reduce(
    (parsed, symbol) => (symbol === "->" ? parsed : append(symbol, parsed)),
    []
  ),
  (s) => ({ params: dropLast(1, s), returnType: last(s) })
);

const typeToValue = (type) => {
  if (type === "String") {
    return "stringValue";
  }

  return undefined;
};

export const typeMatch = (type, value) => {
  if (type === PRIMITIVE_TYPES.Boolean) {
    return typeof value === "boolean";
  } else if (type === PRIMITIVE_TYPES.String) {
    return typeof value === "string";
  }
  return false;
};

export const validate = (params) =>
  pipe(
    map((param) => {
      const code = print(param.ast).code;
      const signature = parseSignature(param.signature);

      const withCall = `
        ${code}
        ${param.name}
      `;

      const script = new vm.Script(withCall);
      const fn = curry(script.runInThisContext());

      const output = reduce(
        (returned, param) => returned(typeToValue(param)),
        fn
      )(signature.params);

      if (typeMatch(signature.returnType, output)) {
        return { error: false };
      } else {
        return { error: true, context: {} };
      }
    }),
    reject(propEq("error", false)),
    ifElse(length, F.reject, always(F.resolve(true)))
  )(params);
