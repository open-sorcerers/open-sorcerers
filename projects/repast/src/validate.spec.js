import { readFile } from "torpor";
import { __, chain, pipe, map } from "ramda";
import { fork } from "fluture";
import { parseWithConfig } from "./parser";
import { PRIMITIVE_TYPES } from "./constants";
import { validate, typeMatch } from "./validate";

test("validate", (done) => {
  pipe(
    readFile(__, "utf8"),
    map(parseWithConfig({})),
    chain(validate),
    fork((errors) => {
      expect(errors).toMatchSnapshot();
      done();
    })(done)
  )("./fixture/validation-cases.js");
});

const typeMatchCases = [
  [PRIMITIVE_TYPES.Boolean, true, true],
  [PRIMITIVE_TYPES.Boolean, false, true],
  [PRIMITIVE_TYPES.Boolean, "false", false],
  [PRIMITIVE_TYPES.String, "string", true],
];

typeMatchCases.forEach(([type, value, expected]) => {
  test(`typeMatch should return "${expected}" for type "${type}" and value "${value}"`, () => {
    expect(typeMatch(type, value)).toBe(expected);
  });
});
