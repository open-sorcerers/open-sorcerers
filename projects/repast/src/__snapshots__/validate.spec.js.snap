// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate 1`] = `
Array [
  Object {
    "context": Object {},
    "error": true,
  },
]
`;
